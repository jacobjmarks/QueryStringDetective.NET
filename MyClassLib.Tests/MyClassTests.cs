using System.Text.Json;
using Snapshooter;

namespace MyClassLib.Tests;

public class MyClassTests
{
    [Theory]
    [InlineData("foo")]
    [InlineData("a&q=b&q=c")]
    [InlineData("null")]
    [InlineData("0")]
    [InlineData("1")]
    [InlineData("-1")]
    [InlineData("1&q=2&q=3")]
    [InlineData("1&q=2&q=.3")]
    [InlineData("-1&q=0&q=1")]
    [InlineData("2147483647")]
    [InlineData("-2147483648")]
    [InlineData("9223372036854775807")]
    [InlineData("-9223372036854775808")]
    [InlineData("18446744073709551615")]
    public async Task AssertBinding(string queryString)
    {
        var results = await MyClass.TestQueryStringBindingAsync("?q=" + queryString);

        var content = "Input: " + queryString
            + "\n\nBindings:\n" + string.Join("\n", results.Where(r => !r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Result}"))
            + "\n\nErroneous:\n" + string.Join("\n", results.Where(r => r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Error!.Message}"));

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}