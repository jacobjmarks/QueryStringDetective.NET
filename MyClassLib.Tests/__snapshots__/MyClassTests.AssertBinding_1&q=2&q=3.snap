Input: 1&q=2&q=3

Bindings:
string     "1,2,3"
string[]   ["1","2","3"]
char[]     ["1","2","3"]
int[]      [1,2,3]
double     123
double[]   [1,2,3]
float      123
float[]    [1,2,3]
decimal    123
decimal[]  [1,2,3]
uint[]     [1,2,3]
long[]     [1,2,3]
ulong[]    [1,2,3]
sbyte[]    [1,2,3]
short[]    [1,2,3]
ushort[]   [1,2,3]
byte[]     "AQID"

Erroneous:
char       400 Bad Request
int        400 Bad Request
bool       400 Bad Request
bool[]     400 Bad Request
uint       400 Bad Request
long       400 Bad Request
ulong      400 Bad Request
sbyte      400 Bad Request
short      400 Bad Request
ushort     400 Bad Request
byte       400 Bad Request
